sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1825,18 +1825,21 @@ pi_result piDevicePartition(pi_device Device,
   }
 
   try {
+    pi_platform Platform = Device->Platform;
     auto ZeSubdevices = new ze_device_handle_t[Count];
     ZE_CALL(zeDeviceGetSubDevices(Device->ZeDevice, &Count, ZeSubdevices));
 
     // Wrap the Level Zero sub-devices into PI sub-devices, and write them out.
     for (uint32_t I = 0; I < Count; ++I) {
-      OutDevices[I] = new _pi_device(ZeSubdevices[I], Device->Platform,
-                                     true /* isSubDevice */);
-      pi_result Result = OutDevices[I]->initialize();
+      std::unique_ptr<_pi_device> PiSubDevice(
+          new _pi_device(ZeSubdevices[I], Platform));
+      pi_result Result = PiSubDevice->initialize();
       if (Result != PI_SUCCESS) {
         delete[] ZeSubdevices;
         return Result;
       }
+      OutDevices[I] = PiSubDevice.get();
+      Platform->PiDevicesCache.push_back(std::move(PiSubDevice));
```

Comment:
You are going to add "new" sub-devices every time piDevicePartition is called, but the idea is that this should only be done once (analogous to how we add root-devices only once to the cache). NOTE: every call to zeDeviceGetSubDevices
 will return the same Level Zero handles.

Reply:
Yes, the current piDevicePartition already behaved like that.
We already created a new _pi_device objects for all returned sub-devices from zeDeviceGetSubDevices().
I was only adding the caching capability to this existing behavior.
Do you want to change the current behavior to check the cache before we call zeDeviceGetSubDevices() in addition to caching?
If so, please advise me how to handle multiple levels of partitioning.
AFAIK, a device can be partitioned into 2 subdevices first and then another level of partitioning will partition them into 4 subdevices. 
How can we handle them? Any hint would be appreciated.

Reply:
> AFAIK, a device can be partitioned into 2 subdevices first and then another level of partitioning will partition them into 4 subdevices.

Let's not handle this yet (not yet truly supported scenario). You can just add an assertion that the input device is a root-device.